pkg/alertmanager/amcfg.go

```diff
@@ -0,0 +1,499 @@
+
+		// add receivers to baseConfig.Receivers
+		for _, receiver := range amConfigs[amConfigIdentifier].Spec.Receivers {
+			baseConfig.Receivers = append(baseConfig.Receivers, cg.convertReceiver(&receiver, crKey))
+		}
+
+		// add inhibitRules to baseConfig.InhibitRules
+		for _, inhibitRule := range amConfigs[amConfigIdentifier].Spec.InhibitRules {
+			baseConfig.InhibitRules = append(baseConfig.InhibitRules, convertInhibitRule(&inhibitRule))
+		}
+	}
+
+	// Append subroutes from base to the end, then replace with the new slice
+	subRoutes = append(subRoutes, baseConfig.Route.Routes...)
+	baseConfig.Route.Routes = subRoutes
+
+	return yaml.Marshal(baseConfig)
+}
+
+func convertRoute(in *monitoringv1.Route, crKey types.NamespacedName, firstLevelRoute bool) *route {
+
+	// Enforce continue to be true for main Route in a CR
+	cont := in.Continue
+	if firstLevelRoute {
+		cont = true
+	}
+
+	match := map[string]string{}
+	matchRE := map[string]string{}
+	if firstLevelRoute {
+		match["namespace"] = crKey.Namespace
+	} else {
+		for _, matcher := range in.Matchers {
+			if *matcher.Regex {
+				matchRE[matcher.Name] = matcher.Value
+			} else {
+				match[matcher.Name] = matcher.Value
+			}
+		}
+	}
+	if len(match) == 0 {
+		match = nil
+	}
+	if len(matchRE) == 0 {
+		matchRE = nil
+	}
+
+	var routes []*route = nil
+	if len(in.Routes) > 0 {
+		routes := make([]*route, len(in.Routes))
+		for i := range in.Routes {
+			routes[i] = convertRoute(&in.Routes[i], crKey, true)
+		}
+	}
+
+	receiver := prefixReceiverName(in.Receiver, crKey)
+
+	return &route{
+		Receiver:       receiver,
+		GroupByStr:     in.GroupBy,
+		GroupWait:      in.GroupWait,
+		GroupInterval:  in.GroupInterval,
+		RepeatInterval: in.RepeatInterval,
+		Continue:       cont,
+		Match:          match,
+		MatchRE:        matchRE,
+		Routes:         routes,
+	}
+}
+
+func (cg *configGenerator) convertReceiver(in *monitoringv1.Receiver, crKey types.NamespacedName) *receiver {
+
+	var pagerdutyConfigs []*pagerdutyConfig
+	if l := len(in.PagerDutyConfigs); l > 0 {
+		pagerdutyConfigs = make([]*pagerdutyConfig, l)
+		for i := range in.PagerDutyConfigs {
+			pagerdutyConfigs[i] = cg.convertPagerdutyConfig(in.PagerDutyConfigs[i], crKey)
+		}
+	}
+
+	return &receiver{
+		Name:             prefixReceiverName(in.Name, crKey),
+		PagerdutyConfigs: pagerdutyConfigs,
+	}
+}
+
+func (cg *configGenerator) convertPagerdutyConfig(in monitoringv1.PagerDutyConfig, crKey types.NamespacedName) *pagerdutyConfig {
+
+	sClient := cg.kclient.CoreV1().Secrets(crKey.Namespace)
+
+	out := &pagerdutyConfig{}
+
+	if in.SendResolved != nil {
+		out.VSendResolved = *in.SendResolved
+	}
+
+	if in.RoutingKey != nil {
+		routingKey, err := getCredFromSecret(sClient, *in.RoutingKey, "routing key", crKey.Namespace+"/"+in.RoutingKey.Name, cg.nsSecretCache)
+		if err != nil {
+			level.Warn(cg.logger).Log("msg", "failed to get routing key from secret for pagerduty, ignoring it and continuing",
```

Comment:
Here, and in some other places in this file, if there's an optional field that specifieds a SecretKeySelector or ConfigMapKeySelector that causes an error when we try to get the value, I've just added a warning log and continued. I'm not sure if that's the best approach or not. Suggestions?

Reply:
We've improved things lately on the service monitor side: the reconciliation loop pulls all data from secrets & configmaps before passing it to the code responsible for generating the configuration.
See https://github.com/prometheus-operator/prometheus-operator/blob/master/pkg/prometheus/asset_store.go and https://github.com/prometheus-operator/prometheus-operator/blob/5ab8453b136d58500e8eac9ee5dde149dfdba658/pkg/prometheus/operator.go#L1234

Reply:
Done.